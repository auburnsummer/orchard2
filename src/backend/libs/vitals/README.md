# vitals

Extract metadata from a rdzip.

# Background

vitals grabs metadata using "facets", which are functions defined in the [facets directory](./facets).
Each facet function is responsible for a little bit of metadata. vitals calls these facet functions and builds
a final metadata object out of the result.

An important note is that vitals is deterministic based on the rdlevel. That is, it will always give the exact
same result if you give it the same rdzip. This means that all information must originate from the rdzip itself,
so for instance vitals determines the created date by the created date on the zip file, and not when it reads the file.

Facet functions have access to the following info:

 - The json representation of the rdlevel created by [rdlevel_parse](../rdlevel_parse/README.md)
 - The zipfile (a python [`zipfile.ZipFile`](https://docs.python.org/3/library/zipfile.html) object)
 - The file, as bytes
 - An optional TOML object which may be in the level

The TOML object is parsed based on the first Comment event in the rdlevel that has the text `#orchard` in it. It's
mainly used for overriding metadata where the facet heuristics may be incorrect. I haven't publicised this
functionality yet, so no levels use it as of now.