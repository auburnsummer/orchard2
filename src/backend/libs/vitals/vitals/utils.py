def dig(dig_list, obj):
    "get something from a json dictionary, array, etc by a list of properties"
    final = obj
    for prop in dig_list:
        final = final[prop]
    return final


def try_dig(dig_list, obj):
    "similar to dig() but returns None if prop does not exist"
    if obj is None:
        return None
    try:
        return dig(dig_list, obj)
    except KeyError:
        return None
