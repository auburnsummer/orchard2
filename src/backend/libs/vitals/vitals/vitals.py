import pprint
import sys
import zipfile
from typing import BinaryIO

import toml

from rdlevel_parse import parse
from vitals.facets.artist_list_facet import artist_list_facet
from vitals.facets.rdlevel_sha1_facet import rdlevel_sha1_facet
from vitals.facets.sha1_facet import sha1_facet
from vitals.facets.author_facet import author_facet
from vitals.facets.event_type_facet import event_type_facet
from vitals.facets.bpm_facet import bpm_facet
from vitals.facets.difficulty_facet import difficulty_facet
from vitals.facets.icon_facet import icon_facet
from vitals.facets.key_facet import make_key_facet, make_color_enabled_key_facet
from vitals.facets.player_facet import player_facet
from vitals.facets.tags_facet import tags_facet
from vitals.facets.thumbnail_facet import thumbnail_facet
from vitals.facets.updated_facet import updated_facet

from .pydantic_model import VitalsLevel


class VitalsException(Exception):
    pass


def main(f: BinaryIO):
    facets = {
        ("artist",): make_color_enabled_key_facet(["settings", "artist"]),
        "artist_tokens": artist_list_facet,
        ("song", "song_ct"): make_color_enabled_key_facet(["settings", "song"]),
        "seizure_warning": make_key_facet(["settings", "seizureWarning"], True),
        ("description", "description_ct"): make_color_enabled_key_facet(
            ["settings", "description"]
        ),
        "hue": make_key_facet(["settings", "songNameHue"], 0.0),
        ("authors", "authors_raw"): author_facet,
        ("max_bpm", "min_bpm"): bpm_facet,
        "difficulty": difficulty_facet,
        ("single_player", "two_player"): player_facet,
        "last_updated": updated_facet,
        "tags": tags_facet,
        ("image", "thumb"): thumbnail_facet,
        "icon": icon_facet,
        (
            "has_classics",
            "has_oneshots",
            "has_squareshots",
            "has_freezeshots",
            "has_freetimes",
            "has_holds",
            "has_skipshots",
            "has_window_dance",
        ): event_type_facet,
        "sha1": sha1_facet,
        "rdlevel_sha1": rdlevel_sha1_facet,
    }

    try:
        with zipfile.ZipFile(f) as z:
            with z.open("main.rdlevel", "r") as rdlevel:
                text = rdlevel.read().decode("utf-8-sig")
                parsed = parse(text)

                # get the TOML comment if there is one.
                # there can be only one
                comments = [evt for evt in parsed["events"] if evt["type"] == "Comment"]
                toml_comment = None
                for comment in comments:
                    try:
                        content = comment["text"]
                        if "#orchard" not in content:
                            continue
                        toml_comment = toml.loads(content)
                        break  # only consider the first valid comment we find.
                    except:
                        # it's fine if there isn't a comment.
                        continue

                final = {}
                for key, func in facets.items():
                    try:
                        result = func(
                            **{"obj": parsed, "zip": z, "file": f, "toml": toml_comment}
                        )
                        if isinstance(key, tuple):
                            # multiple keys with (expected) multiple values that map to the keys.
                            for k, v in zip(key, result):
                                final[k] = v
                        else:
                            final[key] = result
                    except Exception as e:
                        raise VitalsException(
                            f"vitals: An unhandled error occured in a facet: {e}"
                        )

                # return final
                return VitalsLevel(**final)

    except zipfile.BadZipFile:
        raise VitalsException(
            "vitals: this is not a zip file, or we couldn't decode it for some reason."
        )
    except KeyError:
        raise VitalsException("vitals: there is no main.rdlevel in the zip.")
