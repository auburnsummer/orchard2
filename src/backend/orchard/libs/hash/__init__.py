from .hash import checksum, sha1, sha256