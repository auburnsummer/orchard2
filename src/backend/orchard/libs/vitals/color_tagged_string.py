# parse a unity-style <color> string
# returns a tuple of the string _without_ the colors, then what bits are colored by 0-indexed ranges.
import re
import msgspec

from typing import List

class InvalidColorException(Exception):
    pass


# (<color=([a-zA-Z0-9#]+)>)|(<\/color>)
TAG_FINDER = re.compile(r"(<color=([a-zA-Z0-9#]+)>)|(</color>)")

class ColorToken(msgspec.Struct):
    len: int
    color: str  # not always a hex literal, this can be a color literal like "red" etc

def parse_color_tagged_string(s: str):
    matches = list(TAG_FINDER.finditer(s.strip()))

    color_stack = ["default"]

    # where we are in the original string
    curr_pos = 0

    # where we are in the string when it has been stripped of color tags.
    literals: List[str] = []
    tokens: List[ColorToken] = []

    for match in matches:
        # from curr_pos to the starting position of this match is a literal.
        literal = s[curr_pos : match.start()]
        literals.append(literal)
        # ...which is whatever color is previous (at the top of the stack.)
        next_token = ColorToken(len=len(literal), color=color_stack[-1])
        tokens.append(next_token)

        opening, color, closing = match.groups()
        if opening:
            # it's an opening tag, so add a new color to the stack.
            if not color:
                raise InvalidColorException(
                    "Opening color tag found without a given color."
                )
            color_stack.append(color)
        if closing:
            color_stack.pop()

        curr_pos = match.end()

    # we might have leftover text.
    leftover = s[curr_pos:]
    if leftover:
        # it's whatever color is left.
        literals.append(leftover)
        next_token = ColorToken(len=len(leftover), color=color_stack[-1])

        tokens.append(next_token)

    tokens = [t for t in tokens if t.len > 0]

    return "".join(literals), tokens
