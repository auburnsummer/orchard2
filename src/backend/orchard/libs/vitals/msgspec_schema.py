from typing import List, Optional, Annotated
from datetime import datetime

import msgspec

class VitalsLevelBaseMutable(msgspec.Struct):
    "The properties on VitalsLevel where it might make sense for the client to be changing them"
    artist: str
    artist_tokens: Annotated[List[str], msgspec.Meta(min_length=1)]
    song: Annotated[str, msgspec.Meta(min_length=1)]
    seizure_warning: bool
    description: str
    hue: float
    authors: Annotated[List[str], msgspec.Meta(min_length=1)]
    authors_raw: str
    max_bpm: float
    min_bpm: float
    difficulty: int
    single_player: bool
    two_player: bool
    last_updated: datetime
    tags: List[str]
    has_classics: bool
    has_oneshots: bool
    has_squareshots: bool
    has_freezeshots: bool
    has_freetimes: bool
    has_holds: bool
    has_skipshots: bool
    has_window_dance: bool

class VitalsLevelBase(VitalsLevelBaseMutable):
    "VitalsLevelBaseMutable + immutable properties"
    sha1: str
    rdlevel_sha1: str
    is_animated: bool
    
class VitalsLevel(VitalsLevelBase):
    "VitalsLevelBaseMutable + VitalsLevelBase + image data"
    image: bytearray
    thumb: bytearray
    icon: Optional[bytearray]