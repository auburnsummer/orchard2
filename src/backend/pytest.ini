[pytest]
addopts = --disable-socket --allow-hosts=testserver,127.0.0.1