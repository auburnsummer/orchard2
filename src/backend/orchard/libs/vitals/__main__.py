import sys
import pprint
from .vitals import main

import msgspec


try:
    file_name = sys.argv[1]
    with open(file_name, "rb") as f:
        vit = main(f)
        with open("./image.webp", "wb") as f:
            f.write(vit.image)

        with open("./thumb.webp", "wb") as f:
            f.write(vit.thumb);

        vit.image = ""
        vit.thumb = ""
        vit.icon = ""

        pprint.pprint(vit)
except IndexError:
    print("Did you pass a file name in?")
