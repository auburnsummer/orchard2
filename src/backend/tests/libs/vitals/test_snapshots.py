import pytest
from orchard.libs.vitals import analyze
from syrupy.filters import props

# for vitals, we just test with snapshots.
# all levels used for snapshots are in the "fixtures" directory.

# snapshots don't necessarily show correctness of the data, just that it hasn't changed.
# so I have to remember to check the data added to the snapshot when I add new things to vitals.

EXCLUDE_FROM_SNAPSHOT = ["image", "thumb", "icon"]


def test_snapshot(rdzip, snapshot, rdzip_path_map):
    with open(rdzip_path_map[rdzip], "rb") as f:
        actual = analyze(f).model_dump(exclude=EXCLUDE_FROM_SNAPSHOT)
        assert actual == snapshot
