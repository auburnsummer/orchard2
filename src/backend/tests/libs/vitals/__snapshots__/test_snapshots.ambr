# serializer version: 1
# name: test_snapshot[Samario_Making_A_Mistake_-_One_Uncued_Tresillo2.rdzip]
  VitalsLevel(artist='Samario (Making A Mistake)', artist_tokens=['Samario'], song='One Uncued Tresillo', seizure_warning=False, description='One Uncued Tresillo\n(song edited from [...]Machinegun Psystyle[...], by Camellia)', hue=0.0, authors=['Samario (Making A Mistake)'], authors_raw='Samario (Making A Mistake)', max_bpm=70.0, min_bpm=40.0, difficulty=0, single_player=True, two_player=True, last_updated=datetime.datetime(2022, 6, 27, 21, 27, 12), tags=['uncued tresillo', 'samario please stop doing this'], has_classics=False, has_oneshots=False, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='495bd9d1b759db53e600b3986df55bdf3a9c3653', rdlevel_sha1='c467e396d29bd65f4a9c8e9abad64b05a4e39600', is_animated=False, image='96898f86831954f3ab2b102028a6ff6af282f6c9', thumb='9252ee4106a46932541fb677cdc8195ec511fa7b', icon='3ce5397b40ff7c63b294a5517144773961c1b7ec')
# ---
# name: test_snapshot[auburnsummer_-_Triplet_Testaaaaa1.rdzip]
  VitalsLevel(artist='auburnsummer', artist_tokens=['auburnsummer'], song='Triplet Test', seizure_warning=False, description='A two-handed level. The eloquent blend between thirds and halves.', hue=0.66, authors=['auburnsummer', 'Xeno'], authors_raw='auburnsummer, Xeno', max_bpm=110.0, min_bpm=110.0, difficulty=2, single_player=True, two_player=False, last_updated=datetime.datetime(2022, 6, 27, 23, 25, 28), tags=['Piano', '2p', 'Two-handed'], has_classics=True, has_oneshots=True, has_squareshots=True, has_freezeshots=False, has_freetimes=False, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='460807d4b87deb9d94ab31369985ba96b5c95291', rdlevel_sha1='077671a2053d503850d2df220c10f7b901ae0963', is_animated=False, image='7ab14161818970106367ece4ad4c6d352a949ab6', thumb='33346c7075c0999edc06c0215e6663862892f7ba', icon='fff61bf5ce1530fb7bb02518065e4473d01a1fd1')
# ---
# name: test_snapshot[banas-QbMoyaG2nfK.rdzip]
  VitalsLevel(artist='placeboing', artist_tokens=['placeboing'], song='BANAS', seizure_warning=False, description='This level works under the assumption that you know how 8beat, 6beat, and Temporal Shift works.\n\nHope you all have a wonderful April Fools Day!\n\nHollaback Girl by Gwen Stefani x Donkey Kong Country 2 OST\n\nBanana CC by Pyre\nBanana Tile by Het', hue=0.17, authors=["April's Only Fool"], authors_raw="April's Only Fool", max_bpm=147.0, min_bpm=143.5, difficulty=1, single_player=True, two_player=False, last_updated=datetime.datetime(2022, 4, 9, 16, 40, 34), tags=['April Fools', 'banana', 'banas', 'temporal shift', '8beat', '6beat'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=True, has_window_dance=False, sha1='862f1ad3d2f6ae0580055700d79b8ae30c8771b2', rdlevel_sha1='cb674be3db4efc34f5cd77ba5138a7bf2fa7ee12', is_animated=False, image='a99c22887f9f170f6feef2284749d0ae79520c5c', thumb='7c54c09dba5cd12e5679897b966084046e08f2f9', icon='caa16546d32e2e80b2db0818e57528d1115b677f')
# ---
# name: test_snapshot[ennui-og-3VFE7LSmzSo.rdzip]
  VitalsLevel(artist='Pseuu', artist_tokens=['Pseuu'], song='ennui.ogg', seizure_warning=False, description='uh... no comment. level made in about 3-4 working hours\n\n- tricky oneshot patterns\n- skipshots\n- holds', hue=0.42, authors=['Pseuu'], authors_raw='Pseuu', max_bpm=126.0, min_bpm=126.0, difficulty=2, single_player=True, two_player=True, last_updated=datetime.datetime(2022, 1, 3, 23, 19, 26), tags=['chiptune', 'short', '2p', 'oneshots', 'skipshots', 'holds'], has_classics=False, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=True, has_skipshots=True, has_window_dance=True, sha1='250da27a5bff4285dcbf22f13c9c287f7eaf30d7', rdlevel_sha1='34e9ce56866b66dec453d148f8c5d4c2bb7c7199', is_animated=False, image='f1a8e97aa2bf633b58b2cbb332d45cdc81060b1f', thumb='dff39e045d4ce54be953b10bba0238ae08d03871', icon='9cc71a984e73f85c24ca8c16a70530f30851fdca')
# ---
# name: test_snapshot[jimmy-s-Ccby1RK3XJK.rdzip]
  VitalsLevel(artist='toonlink, Kenichi Nishikami, Takeru Kanazaki', artist_tokens=['toonlink', 'Kenichi Nishikami', 'Takeru Kanazaki'], song="Jimmy's Aesthetics", seizure_warning=False, description='A neat song from WarioWare DIY', hue=0.19, authors=['Szprycha'], authors_raw='Szprycha', max_bpm=125.0, min_bpm=125.0, difficulty=1, single_player=True, two_player=False, last_updated=datetime.datetime(2020, 6, 20, 23, 9, 28), tags=['siivagunner', 'nintendo', 'warioware', 'body rock'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=False, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='5bb2d7298237d15b44fbb0a9ca963c00ecb58866', rdlevel_sha1='0d2a9edca604c3f965347ef4b840bb43d94d3a76', is_animated=False, image='6b47ab0046ac30d987a82b2d9debb3478d70c696', thumb='70c1630fd9f148a3c5d4fa26e741ece87d928b22', icon='42fea0c4a11cdf265da96f91eca808ff0c59b878')
# ---
# name: test_snapshot[know-you-2zRJRCjPqJx.rdzip]
  VitalsLevel(artist='Andrew Huang', artist_tokens=['Andrew Huang'], song='Know You', seizure_warning=False, description='Two lonely stars, wandering in the dark...', hue=0.47, authors=['lugi'], authors_raw='lugi', max_bpm=78.5, min_bpm=78.5, difficulty=1, single_player=True, two_player=True, last_updated=datetime.datetime(2020, 6, 24, 15, 6, 48), tags=['slow', 'pop'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='f2ad0f76bbfedf96bec3451eb9ead014accb3047', rdlevel_sha1='f67aa02893fdb5ccfad0e445486bd6595801ecdc', is_animated=False, image='a74fcbd3033b7736d483bfb077897c49e369b889', thumb='dcaf5b41805a80b996c0121d335ebdb70a7450a8', icon='d01dbe3bfb5c3036dfe6ebcc9b29b497765c9749')
# ---
# name: test_snapshot[one-forg-NqVywpSrZs1.rdzip]
  VitalsLevel(artist='Quree', artist_tokens=['Quree'], song='One Forgotten Night', seizure_warning=False, description="Patient experiencing sudden heart strain again. Solve the patient's concerns.", hue=0.1, authors=['BluPen'], authors_raw='BluPen', max_bpm=173.0, min_bpm=100.0, difficulty=2, single_player=True, two_player=True, last_updated=datetime.datetime(2022, 8, 19, 21, 53, 2), tags=['quree', 'window dance'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=False, has_holds=False, has_skipshots=False, has_window_dance=True, sha1='a75d03f274c9032a6f75fa32ac57005615c04b47', rdlevel_sha1='f057987ce0007754050d661648112b78b2b6a535', is_animated=False, image='6cb7f5a64535d38051b0e2a7da103f92216aa21e', thumb='23c35badedcd4a479b126623b8bbc40775826c09', icon='96c50b06b9862d0baf8e5cd5cbcb6325d6b06a21')
# ---
# name: test_snapshot[one-uncu-eqSkB1ZX7Xz.rdzip]
  VitalsLevel(artist='Samario (Making A Mistake)', artist_tokens=['Samario (Making A Mistake)'], song='One Uncued Tresillo', seizure_warning=False, description='One Uncued Tresillo\n(song edited from [...]Machinegun Psystyle[...], by Camellia)', hue=0.0, authors=['Samario (Making A Mistake)'], authors_raw='Samario (Making A Mistake)', max_bpm=160.0, min_bpm=160.0, difficulty=0, single_player=True, two_player=True, last_updated=datetime.datetime(2021, 1, 2, 3, 52, 10), tags=['uncued tresillo', 'samario please stop doing this'], has_classics=False, has_oneshots=False, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='1bf690214be382e78daa98a5c7e73b79a59cf610', rdlevel_sha1='1d74eea124e8c1b0735a9dd79c655589a0b72ed5', is_animated=False, image='96898f86831954f3ab2b102028a6ff6af282f6c9', thumb='9252ee4106a46932541fb677cdc8195ec511fa7b', icon='3ce5397b40ff7c63b294a5517144773961c1b7ec')
# ---
# name: test_snapshot[pull-EpgEYcGYPnU.rdzip]
  VitalsLevel(artist='Hot Freaks', artist_tokens=['Hot Freaks'], song='Pull', seizure_warning=False, description="It's not something easy to admit.", hue=0.88, authors=['Labyrinth'], authors_raw='Labyrinth', max_bpm=89.0, min_bpm=88.0, difficulty=1, single_player=True, two_player=True, last_updated=datetime.datetime(2022, 6, 16, 0, 8, 26), tags=['1p', '2p', 'gimmick'], has_classics=True, has_oneshots=False, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='bdbd0b30d6ea29f8a6d2e8602b8829508dcbfc02', rdlevel_sha1='188fc218f13dec8c67e0de931bd5a073e5d239bf', is_animated=False, image='58ca35fe70eb5b5020be8932a8bce7f67c3f0e9e', thumb='3ab70e85a2103f97481bc5bbcae06c2df0699d26', icon='1bab29b7e4da9a580ae2b02d4bb09895f270c414')
# ---
# name: test_snapshot[rgb-7krFijAGF4D.rdzip]
  VitalsLevel(artist='Green wisteria', artist_tokens=['Green wisteria'], song='RGB', seizure_warning=True, description='(CW: Body Horror)\nA white lily bloomed, a black maw devoured.', hue=0.66, authors=['Delta', 'Xeno'], authors_raw='Delta, Xeno', max_bpm=204.0, min_bpm=100.0, difficulty=2, single_player=True, two_player=False, last_updated=datetime.datetime(2022, 9, 23, 9, 27, 28), tags=['compo19', 'competition 19', 'weird', 'red green blue', 'namine ritsu'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='88f48cdbd6a63773fcf44a38f5ed1ece00f399be', rdlevel_sha1='17a41d4c9b575d3202b8832a9d21c3b873d959e5', is_animated=False, image='ddf66bed33142541fe350d4ee7f16b153748eb08', thumb='efe5ef2d6f797f90f7db0d365904def0a99b1cf3', icon='d08c524f9ce3bece75f53bb8a40280d6f2e3da40')
# ---
# name: test_snapshot[sol-e-ma-dVMtiNt856v.rdzip]
  VitalsLevel(artist='Masssive New Krew', artist_tokens=['Masssive New Krew'], song='Sol e Mar', seizure_warning=False, description='A new patient seems to have come in- Nevermind forget i said that he is an emergency case get here quick!', hue=0.92, authors=['Lyco'], authors_raw='Lyco', max_bpm=150.0, min_bpm=150.0, difficulty=3, single_player=True, two_player=False, last_updated=datetime.datetime(2021, 8, 21, 12, 15, 10), tags=[], has_classics=True, has_oneshots=True, has_squareshots=True, has_freezeshots=False, has_freetimes=False, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='f909b918e876d72fe8c3d385ba03940f85b3b906', rdlevel_sha1='e028099e06643a6ec8ff2e3385f6e9189d134dcf', is_animated=False, image='b007ded2dc60845b9069022be4fad26c088abdfa', thumb='50525c3dc700a4551313f39acf03a706b76cb31e', icon='da39a3ee5e6b4b0d3255bfef95601890afd80709')
# ---
# name: test_snapshot[swing-c2QGYcw8GLq.rdzip]
  VitalsLevel(artist='JustMe', artist_tokens=['JustMe'], song='Swing', seizure_warning=False, description="Printin' twang.", hue=0.08, authors=['Xeno'], authors_raw='Xeno', max_bpm=130.0, min_bpm=130.0, difficulty=1, single_player=True, two_player=True, last_updated=datetime.datetime(2020, 12, 1, 18, 30, 20), tags=['Printing', 'CMYK', 'Swing', 'Bounce', 'syncopation', 'gimmick', '1p', '2p', 'XenOnly'], has_classics=True, has_oneshots=True, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='933bbbbd9c1d79a8c3d2946e5748548a414488ea', rdlevel_sha1='cdbcaadea1a271696931e9badab00fc24c9acd51', is_animated=False, image='6e5e03d1a2ef0ec5ef93b7b5835e8cf9d8dca317', thumb='022940383dcc95c955b4a11effbf5212e3cd2776', icon='94f9e7e0b87d7dfa8492a28ac40a97ee65390582')
# ---
# name: test_snapshot[yi-xiao-Hg3afme47nr.rdzip]
  VitalsLevel(artist='遊助', artist_tokens=['遊助'], song='一笑懸命', seizure_warning=True, description="这次的这个谱我真的花了很长时间，大概有......差不多50个小时吧。歌词因为是CC，只能全用平假名了，看起来麻烦也抱歉了...希望这次能取得比上次更好的结果吧。来来回回也看了很多次，感觉想不到有什么能加的了，就这样交了吧。也希望你也能像歌词一样开开心心的，活着本身没有意义，但就是为了知道是为了什么我们才会活着呀。(I don't know how to translate)", hue=0.17, authors=['来因洛特'], authors_raw='来因洛特', max_bpm=138.0, min_bpm=138.0, difficulty=0, single_player=True, two_player=False, last_updated=datetime.datetime(2022, 8, 27, 11, 35, 12), tags=['easy', 'CNcompo 1', 'relaxing', 'lyrics'], has_classics=True, has_oneshots=False, has_squareshots=False, has_freezeshots=False, has_freetimes=True, has_holds=False, has_skipshots=False, has_window_dance=False, sha1='82dfb357c88904400fbf7eebc2ff2d1820ad9f8f', rdlevel_sha1='1ea9de1156f77d09fb4746c5224d5cc7f82dc0dc', is_animated=False, image='111316cfa917f1b74c0501f871239ed2bf69ef9e', thumb='3f111dc86a13d190bd85d651c6c84096b358f0ba', icon='79608ba56630d97a2b81a9869a167f051e641a58')
# ---
